### PR TITLE
fix ec2 scraper attribute drop without describeTags

### PR DESCRIPTION
**Pull request by akats7**
**Description:** <Describe what has changed.>
Minor fix to update ec2 scraper to apply successfully extracted ec2 attributes (instanceId, cloud region, etc..) even when describeTags return an error (likely due to permissions). Currently if tags are not able to be extracted, Detect returns an err and detectResource does not apply the remaining attributes.

**Link to tracking Issue:** N/A

**Testing:** 
Added tests to test usage of describeTags in detect, made minor modification to detector signature to make it testable.  

**Documentation:**

### CODE REVIEW

File: processor/resourcedetectionprocessor/internal/aws/ec2/ec2.go [hunk 1]
```diff
@@ -88,13 +88,13 @@ func (d *Detector) Detect(ctx context.Context) (resource pcommon.Resource, schem
 		client := getHTTPClientSettings(ctx, d.logger)
 		tags, err := connectAndFetchEc2Tags(meta.Region, meta.InstanceID, d.tagKeyRegexes, client)
 		if err != nil {
-			return res, "", fmt.Errorf("failed fetching ec2 instance tags: %w", err)
```
Comment (Aneurysm9) [inline processor/resourcedetectionprocessor/internal/aws/ec2/ec2.go:91]: Isn't this already emitting the detected resource attributes in addition to the error?  Is the problem that it also needs to emit `conventions.SchemaURL` as the second return parameter, or maybe that the handling of this error at the caller should be adjusted?  The [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md#detecting-resource-information-from-the-environment) is pretty clear that this should be treated as an error:

> Note the failure to detect any resource information MUST NOT be considered an error, whereas an error that occurs during an attempt to detect resource information SHOULD be considered an error.

In any event, tests are required for this change to assert the correct thing is happening and to guard against regression with future changes.

Comment (akats7) [reply processor/resourcedetectionprocessor/internal/aws/ec2/ec2.go:91]: Hey @Aneurysm9, the problem is the way the core resourcedetection component handles the returned resource when an error is returned alongside it, 
`	for _, detector := range p.detectors {
		r, schemaURL, err := detector.Detect(ctx)
		if err != nil {
			p.logger.Warn("failed to detect resource", zap.Error(err))
		} else {
			mergedSchemaURL = MergeSchemaURL(mergedSchemaURL, schemaURL)
			MergeResource(res, r, false)
		}
	}`

If it needs to be an error according to the spec, then I believe we need to change the way returned errors are handled to not ignore the attributes that were returned alongside the error, or alternatively maybe register the describeTags portion as it own detector

Comment (akats7) [reply processor/resourcedetectionprocessor/internal/aws/ec2/ec2.go:91]: And sure thing, I'll add test cases for any implementation we might go with

Comment (Aneurysm9) [reply processor/resourcedetectionprocessor/internal/aws/ec2/ec2.go:91]: I'm okay with moving forward with this change as-is (including tests).  We can re-evaluate the error handling behavior at a later date as I'm not sure there'd really be a functional difference.

**Comment by akats7** [comment]
Will add tests

File: processor/resourcedetectionprocessor/internal/aws/ec2/ec2.go [hunk 2]
```diff
@@ -107,17 +108,19 @@ func getClientConfig(ctx context.Context, logger *zap.Logger) *http.Client {
 	return client
 }
 
-func connectAndFetchEc2Tags(region string, instanceID string, tagKeyRegexes []*regexp.Regexp, client *http.Client) (map[string]string, error) {
+func connectAndFetchEc2Tags(svc ec2iface.EC2API, region string, instanceID string, tagKeyRegexes []*regexp.Regexp, client *http.Client) (map[string]string, error) {
+	if svc != nil {
+		return fetchEC2Tags(svc, instanceID, tagKeyRegexes)
```
Comment (Aneurysm9) [inline processor/resourcedetectionprocessor/internal/aws/ec2/ec2.go:113]: This doesn't seem right.  The connect and fetch steps should probably be separated.  Having this function do two different things depending on whether one of its parameters is nil, and one of them being modifying that input parameter, seems wrong.  There should probably be a separate method to create the EC2 client that can then be used by the existing fetch function.

Comment (akats7) [reply processor/resourcedetectionprocessor/internal/aws/ec2/ec2.go:113]: Yeah I also had some qualms about this, is there a reason that there's a new session being created each time? I wanted to refactor this to create the client when we create the detector but figured there was a timeout or something to consider.

Comment (akats7) [reply processor/resourcedetectionprocessor/internal/aws/ec2/ec2.go:113]: and right even if keeping this structure the call to the second fetchEC2Tags probably should just use a new var like in the original code,
`` 
e = ec2.New(sess)
return fetchEC2Tags(e, instanceID, tagKeyRegexes)
``

**Comment by akats7** [comment]
Can we remove the stale label please @Aneurysm9 @dashpole

**Review by Aneurysm9 — APPROVED** [review]

**Comment by akats7** [comment]
Hey @evan-bradley, are we able to get this merged

**Comment by evan-bradley** [comment]
@akats7 Sorry for the delay. It looks like we are having some Go module issues right now, once those are resolved I'll update the branch and merge or ping you if I need you to update anything.

**Pull request merged**